File: haskell/glue/src/Glue/Lib/Bool/Not.hs
module Glue.Lib.Bool.Not where

import Glue.Eval (Eval)
import Glue.IR (IR (..), isFalsy)

not_ :: IR Eval
not_ = NativeFunc notImpl

notImpl :: IR Eval -> Eval (IR Eval)
notImpl = pure . Bool . isFalsy

File: dart/glue/lib/src/lib/bool/not.dart
import 'package:glue/src/eval.dart';
import 'package:glue/src/ir.dart';

/// Logical not function
/// Mirrors Haskell Glue.Lib.Bool.Not.not_ exactly
final Ir not = IrNativeFunc(notImpl);

/// Logical not implementation
/// Mirrors Haskell Glue.Lib.Bool.Not.notImpl exactly
Eval<Ir> notImpl(Ir arg) => Eval.pure(IrBool(isFalsy(arg)));
